File: apps/daemon/src/dataplane/handshake.rs
use anyhow::{anyhow, Result};
use genehub_proto::{DeviceAuth, InviteAuth, PeerAuth, PeerHello, PeerWelcome, TransportKind};

use crate::channel_auth::{self, SessionKey};
use crate::dataplane::endpoint::PeerAccess;
use crate::state::Shared;
use crate::transport::admission::Admission;

pub struct AcceptedPeer {
    pub welcome: PeerWelcome,
    pub key: SessionKey,
    pub access: PeerAccess,
}

/// Performs one PSK mutual handshake before any business stream exists.
pub fn accept(
    state: &Shared,
    transport: TransportKind,
    admission: Admission,
    hello_wire: &[u8],
    workspace_id: Option<String>,
    workspace_handle: Option<String>,
) -> Result<AcceptedPeer> {
    if hello_wire.is_empty() || hello_wire.len() > genehub_proto::MAX_EXCHANGE_HEAD_BYTES {
        anyhow::bail!("peer hello exceeds its bounded field");
    }
    let hello: PeerHello = serde_json::from_slice(hello_wire)?;
    if hello.version != genehub_proto::DATA_PLANE_VERSION {
        anyhow::bail!(
            "data-plane version mismatch: daemon={}, peer={}",
            genehub_proto::DATA_PLANE_VERSION,
            hello.version
        );
    }
    if hello.client_name.is_empty() || hello.client_name.len() > 80 {
        anyhow::bail!("invalid peer client name");
    }
    let bulk_stream_window = match hello.max_bulk_stream_window_bytes {
        None => genehub_proto::LEGACY_BULK_STREAM_WINDOW_BYTES,
        Some(value)
            if value >= genehub_proto::INITIAL_STREAM_WINDOW_BYTES
                && value <= genehub_proto::MAX_BULK_STREAM_WINDOW_BYTES =>
        {
            value
        }
        Some(_) => anyhow::bail!("invalid peer finite-bulk receive lease"),
    };
    let server_nonce = crate::devices::random_token();
    let (proof, key, device_id, bootstrap_invite) = match (&hello.auth, &admission) {
        (
            PeerAuth::Loopback {
                context,
                nonce,
                proof,
            },
            Admission::Loopback { server_proof },
        ) if transport == TransportKind::Loopback && context == "loopback" => {
            channel_auth::verify_proof(
                &channel_auth::client_proof(server_proof, context, nonce),
                proof,
            )?;
            (
                channel_auth::server_proof(server_proof, context, nonce, &server_nonce),
                channel_auth::derive_key(server_proof, context, nonce, &server_nonce),
                None,
                None,
            )
        }
        (
            PeerAuth::Device {
                device_id,
                nonce,
                proof,
            },
            Admission::DeviceRequired | Admission::Loopback { .. },
        ) => {
            let (id, answer, key) = state.devices.authenticate_session(
                &DeviceAuth {
                    device_id: device_id.clone(),
                    nonce: nonce.clone(),
                    proof: proof.clone(),
                },
                &server_nonce,
            )?;
            (answer, key, Some(id), None)
        }
        (
            PeerAuth::Hosted {
                capability_id,
                nonce,
                proof,
            },
            Admission::Rtc {
                capability_id: expected,
                secret,
                expires_at,
            },
        ) if capability_id == expected && std::time::Instant::now() < *expires_at => {
            let context = channel_auth::hosted_context(capability_id);
            channel_auth::verify_proof(
                &channel_auth::client_proof(secret, &context, nonce),
                proof,
            )?;
            (
                channel_auth::server_proof(secret, &context, nonce, &server_nonce),
                channel_auth::derive_key(secret, &context, nonce, &server_nonce),
                None,
                None,
            )
        }
        (
            PeerAuth::Hosted {
                capability_id,
                nonce,
                proof,
            },
            Admission::Fabric {
                capability_id: expected,
                secret,
                expires_at,
            },
        ) if capability_id == expected && std::time::Instant::now() < *expires_at => {
            let context = channel_auth::hosted_context(capability_id);
            channel_auth::verify_proof(
                &channel_auth::client_proof(secret, &context, nonce),
                proof,
            )?;
            (
                channel_auth::server_proof(secret, &context, nonce, &server_nonce),
                channel_auth::derive_key(secret, &context, nonce, &server_nonce),
                None,
                None,
            )
        }
        (
            PeerAuth::Invite {
                invite_id,
                nonce,
                proof,
            },
            Admission::DeviceRequired | Admission::Loopback { .. },
        ) => {
            let (id, answer, key) = state.devices.authenticate_invite(
                &InviteAuth {
                    invite_id: invite_id.clone(),
                    nonce: nonce.clone(),
                    proof: proof.clone(),
                },
                &server_nonce,
            )?;
            (answer, key, None, Some(id))
        }
        _ => return Err(anyhow!("peer authentication does not match this admission")),
    };

    Ok(AcceptedPeer {
        welcome: PeerWelcome {
            version: genehub_proto::DATA_PLANE_VERSION,
            server_nonce,
            proof,
            max_bulk_stream_window_bytes: Some(bulk_stream_window),
        },
        key,
        access: PeerAccess {
            transport,
            device_id,
            workspace_id,
            workspace_handle,
            bootstrap_invite,
        },
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn loopback_is_a_full_mutual_psk_handshake() {
        let dir = tempfile::tempdir().unwrap();
        let (state, _) = crate::AppState::build(crate::config::Paths::new(dir.path()))
            .await
            .unwrap();
        let secret = "one-use-server-proof";
        let nonce = "00112233445566778899aabbccddeeff";
        let hello = PeerHello {
            version: genehub_proto::DATA_PLANE_VERSION,
            client_name: "test".into(),
            auth: PeerAuth::Loopback {
                context: "loopback".into(),
                nonce: nonce.into(),
                proof: channel_auth::client_proof(secret, "loopback", nonce),
            },
            rtc_supported: true,
            max_bulk_stream_window_bytes: Some(genehub_proto::MAX_BULK_STREAM_WINDOW_BYTES),
        };
        let accepted = accept(
            &state,
            TransportKind::Loopback,
            Admission::Loopback {
                server_proof: secret.into(),
            },
            &serde_json::to_vec(&hello).unwrap(),
            None,
            None,
        )
        .unwrap();
        channel_auth::verify_proof(
            &channel_auth::server_proof(secret, "loopback", nonce, &accepted.welcome.server_nonce),
            &accepted.welcome.proof,
        )
        .unwrap();
        assert_eq!(
            accepted.welcome.max_bulk_stream_window_bytes,
            Some(genehub_proto::MAX_BULK_STREAM_WINDOW_BYTES)
        );
    }

    #[tokio::test]
    async fn a_loopback_listener_also_accepts_a_pairing_invite() {
        let dir = tempfile::tempdir().unwrap();
        let (state, _) = crate::AppState::build(crate::config::Paths::new(dir.path()))
            .await
            .unwrap();
        let invite = state.devices.invite();
        let (invite_id, secret) = invite.code.split_once('.').unwrap();
        let nonce = "00112233445566778899aabbccddeeff";
        let context = format!("invite:{invite_id}");
        let hello = PeerHello {
            version: genehub_proto::DATA_PLANE_VERSION,
            client_name: "pairing".into(),
            auth: PeerAuth::Invite {
                invite_id: invite_id.into(),
                nonce: nonce.into(),
                proof: channel_auth::client_proof(secret, &context, nonce),
            },
            rtc_supported: false,
            max_bulk_stream_window_bytes: None,
        };
        let accepted = accept(
            &state,
            TransportKind::Loopback,
            Admission::Loopback {
                server_proof: "unused-owner-proof".into(),
            },
            &serde_json::to_vec(&hello).unwrap(),
            None,
            None,
        )
        .expect("invite authenticates on the loopback listener");
        channel_auth::verify_proof(
            &channel_auth::server_proof(secret, &context, nonce, &accepted.welcome.server_nonce),
            &accepted.welcome.proof,
        )
        .unwrap();
        assert_eq!(
            accepted.welcome.max_bulk_stream_window_bytes,
            Some(genehub_proto::LEGACY_BULK_STREAM_WINDOW_BYTES)
        );
        assert_eq!(accepted.access.bootstrap_invite.as_deref(), Some(invite_id));
    }
}

File: packages/workbench/src/dataplane/handshake.ts
import type { PeerAuth, PeerHello, PeerWelcome } from "@genehub/proto";

import {
  channelClientProof,
  channelServerProof,
  deriveChannelSessionKey,
  deviceChannelContext,
  hostedChannelContext,
  randomNonce,
  type ChannelSessionKey,
} from "../devices/proof";
import {
  DATA_PLANE_VERSION,
  INITIAL_STREAM_WINDOW_BYTES,
  MAX_BULK_STREAM_WINDOW_BYTES,
} from "./frame";

export type PeerCredential =
  | { kind: "loopback"; secret: string }
  | { kind: "device"; deviceId: string; secret: string }
  | { kind: "hosted"; capabilityId: string; secret: string }
  | { kind: "invite"; inviteId: string; secret: string };

export interface PreparedPeerHandshake {
  hello: PeerHello;
  complete(welcome: PeerWelcome): Promise<PeerHandshakeResult>;
}

export interface PeerHandshakeResult {
  key: ChannelSessionKey;
  /** Receive lease advertised by this authenticated daemon for finite bulk flows. */
  maxBulkStreamWindowBytes: number;
}

/** Builds the only plaintext application message on a peer carrier. */
export async function preparePeerHandshake(
  credential: PeerCredential,
  options: { clientName?: string; rtcSupported?: boolean } = {},
): Promise<PreparedPeerHandshake> {
  const nonce = randomNonce();
  const context = contextOf(credential);
  const proof = await channelClientProof(credential.secret, context, nonce);
  const auth: PeerAuth = authOf(credential, nonce, proof);
  const hello: PeerHello = {
    version: DATA_PLANE_VERSION,
    clientName: boundedClientName(options.clientName),
    auth,
    rtcSupported: options.rtcSupported ?? supportsRtc(),
  };
  return {
    hello,
    async complete(welcome) {
      if (
        welcome.version !== DATA_PLANE_VERSION ||
        typeof welcome.serverNonce !== "string" ||
        !/^[0-9a-f]{32,128}$/.test(welcome.serverNonce) ||
        typeof welcome.proof !== "string"
      ) {
        throw new Error("the daemon returned an invalid data-plane welcome");
      }
      const expected = await channelServerProof(
        credential.secret,
        context,
        nonce,
        welcome.serverNonce,
      );
      if (!sameHex(welcome.proof, expected)) {
        throw new Error("the peer did not prove the expected E2EE secret");
      }
      const advertised = welcome.maxBulkStreamWindowBytes;
      const maxBulkStreamWindowBytes =
        advertised === undefined
          ? INITIAL_STREAM_WINDOW_BYTES
          : validBulkWindow(advertised);
      return {
        key: await deriveChannelSessionKey(
          credential.secret,
          context,
          nonce,
          welcome.serverNonce,
        ),
        maxBulkStreamWindowBytes,
      };
    },
  };
}

function validBulkWindow(value: number): number {
  if (
    !Number.isSafeInteger(value) ||
    value < INITIAL_STREAM_WINDOW_BYTES ||
    value > MAX_BULK_STREAM_WINDOW_BYTES
  ) {
    throw new Error("the daemon advertised an invalid finite-bulk receive lease");
  }
  return value;
}

function contextOf(credential: PeerCredential): string {
  switch (credential.kind) {
    case "loopback":
      return "loopback";
    case "device":
      return deviceChannelContext(credential.deviceId);
    case "hosted":
      return hostedChannelContext(credential.capabilityId);
    case "invite":
      return `invite:${credential.inviteId}`;
  }
}

function authOf(
  credential: PeerCredential,
  nonce: string,
  proof: string,
): PeerAuth {
  switch (credential.kind) {
    case "loopback":
      return { type: "loopback", context: "loopback", nonce, proof };
    case "device":
      return { type: "device", deviceId: credential.deviceId, nonce, proof };
    case "hosted":
      return {
        type: "hosted",
        capabilityId: credential.capabilityId,
        nonce,
        proof,
      };
    case "invite":
      return { type: "invite", inviteId: credential.inviteId, nonce, proof };
  }
}

function boundedClientName(value = "genehub-web"): string {
  const name = value.trim();
  if (!name || new TextEncoder().encode(name).byteLength > 80) {
    throw new Error("invalid data-plane client name");
  }
  return name;
}

function supportsRtc(): boolean {
  return typeof globalThis.RTCPeerConnection === "function";
}

function sameHex(left: string, right: string): boolean {
  if (
    left.length !== right.length ||
    !/^[0-9a-f]+$/i.test(left) ||
    !/^[0-9a-f]+$/i.test(right)
  ) {
    return false;
  }
  let difference = 0;
  for (let index = 0; index < left.length; index += 1) {
    difference |= left.charCodeAt(index) ^ right.charCodeAt(index);
  }
  return difference === 0;
}
